perf(core): add micro benchmark for destroy hook invocation

crisbeto opened the pull request:
Adds a micro benchmark which excercises the logic that invokes `ngOnDestroy` hooks.

crisbeto commented on packages/core/test/render3/perf/view_destroy_hook/index.ts:
```diff
@@ -0,0 +1,27 @@
+// create a view with 1000 sub-views that have onDestroy hooks and destroy them
+console.profile('view_destroy_hook');
+while (refreshTime()) {
+  const rootLView = setupRootViewWithEmbeddedViews(
```
Note: I'm not totally sure whether that this is the correct approach, because it's timing both creation and destruction. Ideally we'd do `while (refreshTime()) destroyViewTree()`, but from what I can tell, that'll destroy the views once and then keep trying to destroy the already-destroyed views. I'm open to suggestions on how to do it better.

pkozlowski-opensource replied:
Yeh, the way I understand the current perf harness is that we don't have a place for prepare / tear-down logic that wouldn't be measured. But I think that it is fine in this particular case as registering destroy hooks takes time as well.

devversion reviewed (approved):
LGTM for dev-infra.

pkozlowski-opensource commented on packages/core/test/render3/perf/setup.ts:
```diff
@@ -79,7 +82,11 @@ export function setupTestHarness(
+    insertView(tView, lView, lContainer, i);
+    if (destroyHook !== null) {
+      (tView.destroyHooks || (tView.destroyHooks = [])).push(i, destroyHook);
+    }
```
Hmmmmm, I think that I would rather prefer that you create a real component / directive with a destroy hook and use it in an embedded view. This way:
* we can test different scenarios easily (directives, directives with providers that have onDestroy hook etc.);
* we don't have to modify `setupRootViewWithEmbeddedViews`;
* we've got less dependencies on the internal impl details (`destroyViewTree`)

WDYT?

pkozlowski-opensource reviewed (commented):
I think that I would rather prefer that you create a real component / directive with a destroy hook and use it in an embedded view, see: https://github.com/angular/angular/pull/35784/files#r386395833

WDYT?

crisbeto committed:
perf(core): add micro benchmark for destroy hook invocation

Adds a micro benchmark which excercises the logic that invokes `ngOnDestroy` hooks.

crisbeto commented:
I've reworked it based on the discussions @pkozlowski-opensource. Can you take another look?

pkozlowski-opensource reviewed (approved):
LGTM, thnx!
